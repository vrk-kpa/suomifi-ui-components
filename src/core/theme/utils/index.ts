export {
  withSuomifiDefaultProps,
  themeOrTokens,
  TokensOrThemeProps,
} from './defaults';
export { focus } from './focus';
export { boxshadowOutline } from './outline';
export { margin, padding, spacingModifiers } from './spacing';
export { typographyUtils, fontSize } from './typography';
