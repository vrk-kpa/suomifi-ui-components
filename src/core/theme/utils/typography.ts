import {
  cssObjectsToCss,
  FlattenSimpleInterpolation,
  cssValueToString,
} from '../../../utils/css';
import { tokens, TypograhpyDesingTokens } from 'suomifi-design-tokens';
import { SuomifiTheme } from '../';

export type TypographyTokensAsCss = {
  [key in keyof typeof tokens.typography]: string
};

type TypographyTokensAsCssProp = {
  [key in keyof TypograhpyDesingTokens]: FlattenSimpleInterpolation
};
export interface TypographyUtil extends TypographyTokensAsCssProp {}
export class TypographyUtil {
  static instance: TypographyUtil;
  constructor(tokens: TypograhpyDesingTokens) {
    // If instance not created, does not take on account if tokens is different!
    if (!TypographyUtil.instance) {
      // Assing typographyTokens as CSS FlattenSimpleInterpolation to this object
      Object.assign(this, cssObjectsToCss(tokens));
      TypographyUtil.instance = this;
    }
    return TypographyUtil.instance;
  }
}

/**
 * Add font-declarations to typography tokens
 * @param typography Typography-tokens
 * @return typographyWithUtils Typography-tokens and typography-tokens as CSS strings
 */
export const typographyUtils = (typography: TypograhpyDesingTokens) => {
  const instance = new TypographyUtil(typography);
  Object.freeze(instance);
  return instance;
};

type ValueTypographyTokenProp = keyof typeof tokens.typography;
export const fontSize = (props: {
  theme: SuomifiTheme;
  [key: string]: any | any[];
}) => (typographyToken: ValueTypographyTokenProp) => {
  return cssValueToString(
    props.theme.values.typography[typographyToken].fontSize,
  );
};
