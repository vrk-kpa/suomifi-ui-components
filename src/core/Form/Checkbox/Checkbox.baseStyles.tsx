import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element, fixInternalMargins, font } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

/* stylelint-disable no-descending-specificity */
const checkedStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--checked {
    & .fi-checkbox_label {
      &::before {
        border-color: ${theme.colors.highlightBase};
        background-color: ${theme.colors.highlightBase};
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.whiteBase};
      }
    }
    &.fi-checkbox--disabled {
      & .fi-checkbox_label {
        &::before {
          border-color: ${theme.colors.depthLight1};
          background-color: ${theme.colors.depthLight1};
        }
        & > .fi-checkbox_icon .fi-icon-base-fill {
          fill: ${theme.colors.depthLight3};
        }
      }
    }
    &.fi-checkbox--error {
      & .fi-checkbox_label {
        &::before {
          border-color: ${theme.colors.alertBase};
          background-color: ${theme.colors.alertBase};
          border-width: 2px;
        }
        & > .fi-checkbox_icon .fi-icon-base-fill {
          fill: ${theme.colors.whiteBase};
        }
      }
    }
  }
`;

const disabledStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--disabled {
    & .fi-checkbox_label {
      cursor: not-allowed;
      color: ${theme.colors.depthBase};
      &::before {
        background-color: ${theme.colors.depthLight3};
        border-color: ${theme.colors.depthLight1};
        border-width: 1px;
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.depthLight3};
      }
    }
    & .fi-hint-text {
      color: ${theme.colors.depthBase};
    }
    &.fi-checkbox--large {
      & .fi-checkbox_label::before {
        border-width: 2px;
      }
    }
  }
`;

const errorStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--error {
    & .fi-checkbox_label {
      &::before {
        border-color: ${theme.colors.alertBase};
        border-width: 2px;
      }
      & > .fi-checkbox_icon .fi-icon-base-fill {
        fill: ${theme.colors.alertBase};
      }
    }
  }
`;

const largeVariantStyles = (theme: SuomifiTheme) => css`
  &.fi-checkbox--large {
    & .fi-checkbox_label {
      padding-left: ${theme.spacing.xxl};
      line-height: 30px;

      &::before {
        content: '';
        position: absolute;
        left: 0px;
        top: 0px;
        box-sizing: border-box;
        height: 30px;
        width: 30px;
        color: ${theme.colors.depthDark3};
        border: 2px solid;
      }
      & .fi-checkbox_icon {
        height: 20px;
        width: 20px;
        left: 5px;
        top: 4px;
      }
    }
    & .fi-hint-text {
      padding-left: ${theme.spacing.xxl};
    }

    &:focus-within {
      & .fi-checkbox_label {
        &::after {
          width: 32px;
          height: 32px;
          top: -1px;
          left: -1px;
        }
      }
    }
  }
`;

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${font(theme)('bodyTextSmall')}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}
  ${fixInternalMargins()}
  
  & .fi-checkbox_label {
    position: relative;
    display: block;
    padding-left: ${theme.spacing.l};
    cursor: pointer;
    line-height: 27px;

    &::before {
      content: '';
      position: absolute;
      left: 0px;
      top: ${theme.spacing.xxs};
      box-sizing: border-box;
      height: 18px;
      width: 18px;
      border: 1px solid ${theme.colors.depthDark3};
      border-radius: ${theme.radiuses.basic};
      background-color: ${theme.colors.whiteBase};
    }
  }

  & .fi-checkbox_icon {
    position: absolute;
    height: 10px;
    width: 10px;
    left: 4px;
    top: 9px;
  }

  &:focus-within {
    & .fi-checkbox_label {
      /* Modified version of theme.focuses.absoluteFocus */
      &::after {
        content: '';
        position: absolute;
        pointer-events: none;
        top: 4px;
        left: -1px;
        border-radius: 2px;
        background-color: transparent;
        border: 0px solid ${theme.colors.whiteBase};
        box-sizing: border-box;
        box-shadow: 0 0 0 2px ${theme.colors.accentSecondary};
        z-index: 9999;
        width: 20px;
        height: 20px;
        outline: 2px solid transparent; /* For high contrast mode */
      }
    }
  }

  & .fi-checkbox_input {
    position: absolute;
    opacity: 0;
    z-index: -9999;
  }

  & .fi-hint-text {
    padding-left: ${theme.spacing.l};
    color: ${theme.colors.depthDark1};
  }

  & .fi-status-text {
    line-height: 18px;
    &.fi-checkbox_statusText--has-content {
      margin-top: ${theme.spacing.xxs};
    }
  }

  ${largeVariantStyles(theme)};
  ${checkedStyles(theme)};
  ${errorStyles(theme)};
  ${disabledStyles(theme)};
`;
