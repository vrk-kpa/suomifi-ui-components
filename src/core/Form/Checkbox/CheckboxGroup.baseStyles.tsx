import { css } from 'styled-components';
import { SuomifiTheme } from '../../theme';
import { element } from '../../theme/reset';
import { MarginProps, buildSpacingCSS } from '../../theme/utils/spacing';

export const baseStyles = (
  theme: SuomifiTheme,
  globalMargins?: MarginProps,
  propMargins?: MarginProps,
) => css`
  ${element(theme)}
  ${buildSpacingCSS(globalMargins)}
  ${buildSpacingCSS(propMargins, true)}

  & .fi-checkbox-group_legend {
    &--with-margin {
      margin-bottom: ${theme.spacing.xs};
    }
    .fi-hint-text {
      margin-bottom: 0;
    }
  }

  & .fi-checkbox-group_label--with-margin {
    margin-bottom: 10px;
  }

  & .fi-checkbox-group_container {
    & > .fi-checkbox {
      margin-bottom: ${theme.spacing.xs};

      &:last-child {
        margin-bottom: 0;
      }
    }

    & .fi-checkbox--large {
      margin-top: ${theme.spacing.xxs};
      margin-bottom: ${theme.spacing.s};
    }
  }
`;
