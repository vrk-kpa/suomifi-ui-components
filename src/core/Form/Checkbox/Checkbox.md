```js
import { Checkbox } from './Checkbox';
import { useState } from 'react';

const [checked, setChecked] = useState(false);
<>
  <Checkbox defaultChecked hintText="This is an example hint text">
    Regular checkbox that is checked and has a hint text
  </Checkbox>

  <Checkbox
    hintText="Example hint text"
    status="error"
    statusText="You need to accept the terms and conditions to continue"
  >
    Regular checkbox with a hint text and an error message
  </Checkbox>

  <Checkbox.large>Large checkbox</Checkbox.large>

  <Checkbox.large defaultChecked hintText="Example hint text">
    Checked large checkbox with a hint text
  </Checkbox.large>

  <Checkbox disabled>Disabled checkbox</Checkbox>

  <Checkbox
    checked={checked}
    onClick={({ checkboxState }) => {
      if (window.confirm('Change checkbox state?')) {
        setChecked(checkboxState);
      }
    }}
  >
    Checkbox with controlled checked state
  </Checkbox>
</>;
```
