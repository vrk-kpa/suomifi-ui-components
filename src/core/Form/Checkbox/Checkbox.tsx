import React, { Component, forwardRef, ReactNode } from 'react';
import { default as styled } from 'styled-components';
import classnames from 'classnames';
import { InputStatus, StatusTextCommonProps } from '../types';
import { getConditionalAriaProp } from '../../../utils/aria';
import { getLogger } from '../../../utils/log';
import { AutoId } from '../../utils/AutoId/AutoId';
import {
  SuomifiThemeProp,
  SuomifiThemeConsumer,
  SpacingConsumer,
} from '../../theme';
import {
  separateMarginProps,
  MarginProps,
  GlobalMarginProps,
} from '../../theme/utils/spacing';
import { HtmlLabel, HtmlDiv, HtmlInput, HtmlInputProps } from '../../../reset';
import { StatusText } from '../StatusText/StatusText';
import { HintText } from '../HintText/HintText';
import { CheckboxGroupConsumer } from './CheckboxGroup';
import { baseStyles } from './Checkbox.baseStyles';
import { IconCheck } from 'suomifi-icons';
import { filterDuplicateKeys } from '../../../utils/common/common';

const baseClassName = 'fi-checkbox';

const checkboxClassNames = {
  container: `${baseClassName}_container`,
  input: `${baseClassName}_input`,
  label: `${baseClassName}_label`,
  statusTextHasContent: `${baseClassName}_statusText--has-content`,
  hintText: `${baseClassName}_hintText`,
  disabled: `${baseClassName}--disabled`,
  error: `${baseClassName}--error`,
  checked: `${baseClassName}--checked`,
  large: `${baseClassName}--large`,
  icon: `${baseClassName}_icon`,
};

type CheckboxStatus = Exclude<InputStatus, 'success'>;

export interface CheckboxProps
  extends StatusTextCommonProps,
    MarginProps,
    Omit<HtmlInputProps, 'onClick' | 'value'> {
  /** Controlled checked state  */
  checked?: boolean;
  /** Default status of Checkbox when not using controlled `checked` state
   * @default false
   */
  defaultChecked?: boolean;
  /** CSS class for custom styles */
  className?: string;
  /** Disables the Checkbox. Value won't be included in a form when submitting */
  disabled?: boolean;
  /** Callback fired on Checkbox click */
  onClick?: ({ checkboxState }: { checkboxState: boolean }) => void;
  /**
   * Children will be rendered as Checkbox's label. Use a short and descriptive label.
   */
  children?: ReactNode;
  /**
   * Variant of the Checkbox
   * @default small
   */
  variant?: 'small' | 'large';
  /**
   * `'default'` | `'error'`
   *
   * Status of the component. Error state creates a red border around the Checkbox.
   * Always use a descriptive `statusText` with an error status.
   * @default default
   */
  status?: CheckboxStatus;
  /**
   * Hint text to be displayed under the label
   */
  hintText?: string;
  /**
   * aria-label for the HTML input element.
   * Alternatively you can define an aria-labelledby
   */
  'aria-label'?: string;
  /**
   * aria-labelledby for the HTML input element.
   * Alternatively you can define an aria-label
   */
  'aria-labelledby'?: string;
  /**
   * aria-describedby for the HTML input element,
   */
  'aria-describedby'?: string;
  /**
   * HTML id attribute.
   * If no id is specified, one will be generated automatically
   */
  id?: string;
  /** HTML name attribute */
  name?: string;
  /** HTML value attribute */
  value?: string;
  /** Ref is passed to the underlying input element. Alternative to React `ref` attribute. */
  forwardedRef?: React.RefObject<HTMLInputElement>;
}

class BaseCheckbox extends Component<CheckboxProps> {
  state = {
    checkedState: !!this.props.checked || !!this.props.defaultChecked,
  };

  static getDerivedStateFromProps(
    nextProps: CheckboxProps,
    prevState: { checkedState: boolean },
  ) {
    const { checked } = nextProps;
    if (checked !== undefined && checked !== prevState.checkedState) {
      return { checkedState: checked };
    }
    return null;
  }

  handleClick = () => {
    const { onClick, checked } = this.props;
    const { checkedState } = this.state;
    if (checked === undefined) {
      this.setState({ checkedState: !checkedState });
    }
    if (!!onClick) {
      onClick({ checkboxState: !checkedState });
    }
  };

  render() {
    const {
      id,
      className,
      disabled = false,
      'aria-label': ariaLabel,
      'aria-labelledby': ariaLabelledBy,
      'aria-describedby': ariaDescribedBy,
      statusTextAriaLiveMode = 'assertive',
      children,
      checked: dismissChecked,
      defaultChecked: dismissDefaultChecked,
      onClick: dismissOnClick,
      hintText,
      status,
      statusText,
      name,
      value,
      forwardedRef,
      onClick,
      variant,
      style,
      ...rest
    } = this.props;
    const [_marginProps, passProps] = separateMarginProps(rest);

    const { checkedState } = this.state;

    if (!children) {
      getLogger().error(
        'Checkbox component should have a label or a child element that acts as one. Add label content or a child element.',
      );
    }

    if (
      ('name' in this.props && (typeof name !== 'string' || name === '')) ||
      ('value' in this.props && (typeof value !== 'string' || value === ''))
    ) {
      getLogger().warn(
        'Name and value props should have non-empty values if provided.',
      );
    }

    const statusTextId = !!statusText ? `${id}-statusText` : undefined;
    const hintTextId = !!hintText ? `${id}-hintText` : undefined;

    return (
      <HtmlDiv
        className={classnames(
          checkboxClassNames.container,
          className,
          baseClassName,
          {
            [checkboxClassNames.error]: status === 'error' && !disabled,
            [checkboxClassNames.checked]: checkedState,
            [checkboxClassNames.large]: variant === 'large',
            [checkboxClassNames.disabled]: !!disabled,
          },
        )}
        style={style}
      >
        <HtmlInput
          type="checkbox"
          disabled={disabled}
          id={id}
          {...getConditionalAriaProp('aria-label', [ariaLabel])}
          {...getConditionalAriaProp('aria-labelledby', [ariaLabelledBy])}
          {...getConditionalAriaProp('aria-describedby', [
            statusTextId,
            hintTextId,
            ariaDescribedBy,
          ])}
          aria-invalid={status === 'error'}
          checked={!!checkedState}
          className={checkboxClassNames.input}
          onChange={this.handleClick}
          name={name}
          forwardedRef={forwardedRef}
          {...(value ? { value } : {})}
          {...passProps}
        />
        <HtmlLabel htmlFor={id} className={checkboxClassNames.label}>
          {!!checkedState && <IconCheck className={checkboxClassNames.icon} />}
          {children}
        </HtmlLabel>
        <HintText id={hintTextId}>{hintText}</HintText>
        <StatusText
          className={classnames({
            [checkboxClassNames.statusTextHasContent]: !!statusText,
          })}
          id={statusTextId}
          status={status}
          disabled={disabled}
          ariaLiveMode={statusTextAriaLiveMode}
        >
          {statusText}
        </StatusText>
      </HtmlDiv>
    );
  }
}

const StyledCheckbox = styled(
  ({
    theme,
    globalMargins,
    ...passProps
  }: CheckboxProps & SuomifiThemeProp & GlobalMarginProps & MarginProps) => (
    <BaseCheckbox {...passProps} />
  ),
)`
  ${({ theme, globalMargins, ...rest }) => {
    const [marginProps, _passProps] = separateMarginProps(rest);
    const cleanedGlobalMargins = filterDuplicateKeys(
      globalMargins.checkbox,
      marginProps,
    );
    return baseStyles(theme, cleanedGlobalMargins, marginProps);
  }}
`;

const Checkbox = forwardRef(
  (props: CheckboxProps, ref: React.RefObject<HTMLInputElement>) => {
    const { id: propId, status: propStatus, ...passProps } = props;
    return (
      <SpacingConsumer>
        {({ margins }) => (
          <SuomifiThemeConsumer>
            {({ suomifiTheme }) => (
              <AutoId id={propId}>
                {(id) => (
                  <CheckboxGroupConsumer>
                    {({ status: groupStatus }) => (
                      <StyledCheckbox
                        theme={suomifiTheme}
                        id={id}
                        forwardedRef={ref}
                        globalMargins={margins}
                        status={!!propStatus ? propStatus : groupStatus}
                        {...passProps}
                      />
                    )}
                  </CheckboxGroupConsumer>
                )}
              </AutoId>
            )}
          </SuomifiThemeConsumer>
        )}
      </SpacingConsumer>
    );
  },
);

Checkbox.displayName = 'Checkbox';
export { Checkbox };
