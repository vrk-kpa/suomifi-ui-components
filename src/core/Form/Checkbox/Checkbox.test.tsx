import React from 'react';
import { axeTest } from '../../../utils/test';
import { render } from '@testing-library/react';
import { Checkbox, CheckboxProps } from './Checkbox';

const BaseCheckbox = (props: CheckboxProps) => {
  const { id, children, ...passProps } = props;
  return (
    <Checkbox id="test" {...passProps}>
      {children || 'Default label'}
    </Checkbox>
  );
};

const RegularTestCheckbox = (
  <BaseCheckbox data-testid="regular_id">Regular</BaseCheckbox>
);

const LargeTestCheckboxWithHintText = (
  <BaseCheckbox variant="large" hintText="Take a hint">
    Large
  </BaseCheckbox>
);

const CheckedLargeTestCheckboxWithError = (
  <BaseCheckbox
    variant="large"
    statusText="EROR EROR"
    defaultChecked
    status="error"
  >
    Large Checked with error
  </BaseCheckbox>
);

const DisabledTestCheckbox = (
  <BaseCheckbox disabled>Regular disabled</BaseCheckbox>
);

describe('accessibility', () => {
  test(
    'RegularTestCheckbox should not have basic accessibility issues',
    axeTest(RegularTestCheckbox),
  );

  test(
    'LargeTestCheckboxWithHintText should not have basic accessibility issues',
    axeTest(LargeTestCheckboxWithHintText),
  );

  test(
    'CheckedLargeTestCheckboxWithError should not have basic accessibility issues',
    axeTest(CheckedLargeTestCheckboxWithError),
  );

  test(
    'DisabledTestCheckbox should not have basic accessibility issues',
    axeTest(DisabledTestCheckbox),
  );
});

describe('props', () => {
  describe('name', () => {
    it('has the given name prop on start and after prop change', () => {
      const { getByRole, rerender } = render(<BaseCheckbox name="magical" />);
      const input = getByRole('checkbox');
      expect(input).toHaveAttribute('name', 'magical');

      rerender(<BaseCheckbox name="unicorn" />);
      expect(input).toHaveAttribute('name', 'unicorn');
    });
  });

  describe('value', () => {
    it('do not have value attribute if prop is not given', () => {
      const { getByRole } = render(<BaseCheckbox />);
      const input = getByRole('checkbox');
      expect(input).not.toHaveAttribute('value');
    });

    it('has the given value prop on start and after prop change', () => {
      const { getByRole, rerender } = render(
        <BaseCheckbox value="test-value-1" />,
      );
      const input = getByRole('checkbox');
      expect(input).toHaveAttribute('value', 'test-value-1');

      rerender(<BaseCheckbox value="test-value-2" />);
      expect(input).toHaveAttribute('value', 'test-value-2');
    });
  });

  describe('children', () => {
    it('has the given children and it changes on prop change', () => {
      const { getByTestId, rerender } = render(RegularTestCheckbox);
      expect(getByTestId('regular_id').nextElementSibling?.textContent).toBe(
        'Regular',
      );

      rerender(
        <BaseCheckbox data-testid="regular_id_changed">
          Regular changed
        </BaseCheckbox>,
      );
      expect(
        getByTestId('regular_id_changed').nextElementSibling?.textContent,
      ).toBe('Regular changed');
    });

    it('has matching snapshot', () => {
      const { container } = render(RegularTestCheckbox);
      expect(container.firstChild).toMatchSnapshot();
    });
  });

  describe('hintText', () => {
    it('has the hint text element', () => {
      const { getByText } = render(
        <Checkbox hintText="Example hint text">Text</Checkbox>,
      );
      expect(getByText('Example hint text')).toHaveClass('fi-hint-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <Checkbox id="123" hintText="Example hint text">
          Text
        </Checkbox>,
      );
      expect(getByRole('checkbox')).toHaveAttribute(
        'aria-describedby',
        '123-hintText',
      );
    });
  });

  describe('statusText', () => {
    it('has the status text element', () => {
      const { getByText } = render(
        <Checkbox statusText="Example status text">Text</Checkbox>,
      );
      const statusText = getByText('Example status text');
      expect(statusText).toHaveClass('fi-status-text');
    });

    it('will be added to input aria-describedby', () => {
      const { getByRole } = render(
        <Checkbox id="123" statusText="Example status text">
          Text
        </Checkbox>,
      );
      expect(getByRole('checkbox')).toHaveAttribute(
        'aria-describedby',
        '123-statusText',
      );
    });
  });

  describe('disabled', () => {
    it('has disabled attribute and classname', () => {
      const { container, getByRole } = render(
        <Checkbox disabled statusText="Example status text">
          Text
        </Checkbox>,
      );
      expect(container.firstChild).toHaveClass('fi-checkbox--disabled');
      expect(getByRole('checkbox')).toHaveAttribute('disabled');
    });
  });

  describe('className', () => {
    it('has the given custom className', () => {
      const { container } = render(
        <Checkbox className="custom-style">Text</Checkbox>,
      );
      expect(container.firstChild).toHaveClass('custom-style');
    });
  });

  describe('margin', () => {
    it('should have margin style from margin prop', () => {
      const { container } = render(<Checkbox margin="xs">Text</Checkbox>);
      expect(container.firstChild).toHaveStyle('margin: 10px');
    });
  });

  it('should have margin style overridden by style prop', async () => {
    const { container } = render(
      <Checkbox margin="xs" style={{ margin: 2 }}>
        Text
      </Checkbox>,
    );
    expect(container.firstChild).toHaveAttribute('style', 'margin: 2px;');
  });
});
