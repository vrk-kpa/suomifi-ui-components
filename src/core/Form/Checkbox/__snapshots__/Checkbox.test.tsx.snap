// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props children has matching snapshot 1`] = `
.c0 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: block;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c0::before,
.c0::after {
  box-sizing: border-box;
}

.c2 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  overflow: visible;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline-block;
  max-width: 100%;
}

.c2::-webkit-input-placeholder {
  color: inherit;
  opacity: 0.54;
}

.c2::before,
.c2::after {
  box-sizing: border-box;
}

.c3 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  max-width: 100%;
}

.c3::before,
.c3::after {
  box-sizing: border-box;
}

.c4 {
  line-height: 1.15;
  -ms-text-size-adjust: 100%;
  -webkit-text-size-adjust: 100%;
  margin: 0;
  padding: 0;
  border: 0;
  box-sizing: border-box;
  font: 100% inherit;
  line-height: 1;
  text-align: left;
  -webkit-text-decoration: none;
  text-decoration: none;
  vertical-align: baseline;
  color: inherit;
  background: none;
  cursor: inherit;
  display: inline;
  max-width: 100%;
  word-wrap: normal;
  word-break: normal;
  white-space: normal;
}

.c4::before,
.c4::after {
  box-sizing: border-box;
}

.c5 {
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 600;
  color: hsl(0,0%,13%);
  font-size: 14px;
  line-height: 20px;
}

.c5.fi-status-text {
  display: block;
}

.c5.fi-status-text.fi-status-text--error {
  color: hsl(3,59%,43%);
}

.c5.fi-status-text .fi-icon {
  vertical-align: middle;
  -webkit-transform: translateY(-0.1em);
  -ms-transform: translateY(-0.1em);
  transform: translateY(-0.1em);
  margin-right: 5px;
}

.c1 {
  color: hsl(0,0%,13%);
  -webkit-letter-spacing: 0;
  -moz-letter-spacing: 0;
  -ms-letter-spacing: 0;
  letter-spacing: 0;
  -webkit-text-decoration: none;
  text-decoration: none;
  word-break: break-word;
  overflow-wrap: break-word;
  -webkit-font-smoothing: antialiased;
  font-family: 'Source Sans Pro','Helvetica Neue','Arial',sans-serif;
  font-size: 16px;
  line-height: 1.5;
  font-weight: 400;
}

.c1 .fi-label,
.c1 .fi-hint-text,
.c1 .fi-status-text {
  margin: 0;
}

.c1 .fi-checkbox_label {
  position: relative;
  display: block;
  padding-left: 25px;
  cursor: pointer;
  line-height: 27px;
}

.c1 .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 5px;
  box-sizing: border-box;
  height: 18px;
  width: 18px;
  border: 1px solid hsl(201,7%,58%);
  border-radius: 2px;
  background-color: hsl(0,0%,100%);
}

.c1 .fi-checkbox_icon {
  position: absolute;
  height: 10px;
  width: 10px;
  left: 4px;
  top: 9px;
}

.c1:focus-within .fi-checkbox_label::after {
  content: '';
  position: absolute;
  pointer-events: none;
  top: 4px;
  left: -1px;
  border-radius: 2px;
  background-color: transparent;
  border: 0px solid hsl(0,0%,100%);
  box-sizing: border-box;
  box-shadow: 0 0 0 2px hsl(196,77%,44%);
  z-index: 9999;
  width: 20px;
  height: 20px;
  outline: 2px solid transparent;
}

.c1 .fi-checkbox_input {
  position: absolute;
  opacity: 0;
  z-index: -9999;
}

.c1 .fi-hint-text {
  padding-left: 25px;
  color: hsl(202,7%,40%);
}

.c1 .fi-status-text {
  line-height: 18px;
}

.c1 .fi-status-text.fi-checkbox_statusText--has-content {
  margin-top: 5px;
}

.c1.fi-checkbox--large .fi-checkbox_label {
  padding-left: 40px;
  line-height: 30px;
}

.c1.fi-checkbox--large .fi-checkbox_label::before {
  content: '';
  position: absolute;
  left: 0px;
  top: 0px;
  box-sizing: border-box;
  height: 30px;
  width: 30px;
  color: hsl(201,7%,58%);
  border: 2px solid;
}

.c1.fi-checkbox--large .fi-checkbox_label .fi-checkbox_icon {
  height: 20px;
  width: 20px;
  left: 5px;
  top: 4px;
}

.c1.fi-checkbox--large .fi-hint-text {
  padding-left: 40px;
}

.c1.fi-checkbox--large:focus-within .fi-checkbox_label::after {
  width: 32px;
  height: 32px;
  top: -1px;
  left: -1px;
}

.c1.fi-checkbox--checked .fi-checkbox_label::before {
  border-color: hsl(212,63%,45%);
  background-color: hsl(212,63%,45%);
}

.c1.fi-checkbox--checked .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(0,0%,100%);
}

.c1.fi-checkbox--checked.fi-checkbox--disabled .fi-checkbox_label::before {
  border-color: hsl(202,7%,80%);
  background-color: hsl(202,7%,80%);
}

.c1.fi-checkbox--checked.fi-checkbox--disabled .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(202,7%,97%);
}

.c1.fi-checkbox--checked.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,43%);
  background-color: hsl(3,59%,43%);
  border-width: 2px;
}

.c1.fi-checkbox--checked.fi-checkbox--error .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(0,0%,100%);
}

.c1.fi-checkbox--error .fi-checkbox_label::before {
  border-color: hsl(3,59%,43%);
  border-width: 2px;
}

.c1.fi-checkbox--error .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(3,59%,43%);
}

.c1.fi-checkbox--disabled .fi-checkbox_label {
  cursor: not-allowed;
  color: hsl(202,7%,67%);
}

.c1.fi-checkbox--disabled .fi-checkbox_label::before {
  background-color: hsl(202,7%,97%);
  border-color: hsl(202,7%,80%);
  border-width: 1px;
}

.c1.fi-checkbox--disabled .fi-checkbox_label > .fi-checkbox_icon .fi-icon-base-fill {
  fill: hsl(202,7%,97%);
}

.c1.fi-checkbox--disabled .fi-hint-text {
  color: hsl(202,7%,67%);
}

.c1.fi-checkbox--disabled.fi-checkbox--large .fi-checkbox_label::before {
  border-width: 2px;
}

<div
  class="c0 fi-checkbox_container c1 fi-checkbox"
>
  <input
    aria-invalid="false"
    class="c2 fi-checkbox_input"
    data-testid="regular_id"
    id="test"
    type="checkbox"
  />
  <label
    class="c3 fi-checkbox_label"
    for="test"
  >
    Regular
  </label>
  <span
    aria-atomic="true"
    aria-live="assertive"
    class="c4 c5 fi-status-text"
  />
</div>
`;
